framework/src/loops/ComputeMortarFunctor.C

```diff
@@ -82,175 +80,219 @@ ComputeMortarFunctor::operator()()
 
   unsigned int num_cached = 0;
 
-  for (MeshBase::const_element_iterator
-           el = _amg.mortarSegmentMesh().active_local_elements_begin(),
-           end_el = _amg.mortarSegmentMesh().active_local_elements_end();
-       el != end_el;
-       ++el)
-  {
-    // Note: this is a mortar segment mesh Elem, *not* an Elem
-    // from the original mesh or the side of an Elem from the
-    // original mesh.
-    const Elem * msm_elem = *el;
+  // Compile map of secondary to msm elements
+  // All this could be computed in AutomaticMortarGeneration, will suffice for now but
+  // in the future may want to move for optimization, especially on non-displaced meshes
+  std::map<const Elem *, std::vector<Elem *>> secondary_elems_to_mortar_segments;
 
-    // Get a reference to the MortarSegmentInfo for this Elem.
+  for (const auto msm_elem : _amg.mortarSegmentMesh().active_local_element_ptr_range())
+  {
     const MortarSegmentInfo & msinfo = _amg.mortarSegmentMeshElemToInfo().at(msm_elem);
```

Comment:
```suggestion
    const MortarSegmentInfo & msinfo = libmesh_map_find(_amg.mortarSegmentMeshElemToInfo(), msm_elem);
```

Reply:
@lindsayad Is this `libmesh_map_find ` safer to call when we are not sure about non-existing key values?

Reply:
yes. It will print a stack trace if the key doesn't exist instead of just quietly throwing an exception. @permcody hates `at` haha